cmd/gpu_plugin/gpu_plugin.go

```diff
@@ -134,6 +134,9 @@ func (dp *devicePlugin) scan() (dpapi.DeviceTree, error) {
 			return nil, errors.Wrap(err, "Can't read device folder")
 		}
 
+		dat, err = ioutil.ReadFile(path.Join(dp.sysfsDir, f.Name(), "device/sriov_numvfs"))
+		isPFwithVFs := (err == nil && strings.TrimSpace(string(dat)) != "0")
```

Comment:
would it be simpler to `filepath.Glob()` all `virtfn*` for that PF?

Reply:
> would it be simpler to `filepath.Glob()` all `virtfn*` for that PF?

While the reverse check (whether given device is VF) is simpler by checking for the single "sriov_physfn" link, I don't think globbing all the "sriov_virfn*" links and counting the resulting array len is any simpler than checking "sriov_numvfs" file. Most[1] GPUs are not PFs, and "numvfs" check should short-circuits faster as that file doesn't exist for VFs.

[1] There's not much point in enabling SR-IOV unless one sets up multiple VFs for PF.

Reply:
> and checking the resulting array len is any simpler than checking "sriov_numvfs" file

OK. my thinking was that it is simpler because there's no need to open/read files, do input sanitation, and type casting to get  the number out.

Reply:
also see the note about [deprecation](https://golang.org/doc/go1.16#ioutil)

Reply:
> also see the note about deprecation

This is not the only thing using "ioutil" in GPU plugin code, and the new functions will be available only with Go v1.16 => only after project minimum Golang version is raised to v1.16 (currently CI seems to use v1.15), there should be PR to update (all) "ioutil" use.


> my thinking was that it is simpler because there's no need to open/read files, do input sanitation, and type casting to get the number out.

Both are two lines of code.  Second line would be slightly shorter due to array check being simpler, but IMHO numvfs check is clearer to user than counting things with Glob.

Even in cases where the file is actually read, it's only 2-4 bytes, and soon os.ReadFile() will match to single kernel syscall: https://lwn.net/Articles/813827/

"Input sanitation" just removes trailing newline, and there's no type casting (it's a content string comparison, not a number comparison), which is less than what glob needs to do.  Glob needs to read the directory contents and do match against the directory entry names (+ alloc space for the matching file names), which without any VFs is: 
```
$ ls /sys/class/drm/card1/device/|wc
     48      48     560
```

Reply:
I think we have enough due diligence done for this :-)